vita3k/host/src/pup.cpp

```diff
@@ -0,0 +1,473 @@
+    auto ci_off = SceControlInfo::Size;
+
+    if (controlinfo_hdr.type == ControlType::DIGEST_SHA256) {
+        filein.seekg(self_hdr.controlinfo_offset + ci_off);
+        ci_off += SceControlInfoDigest256::Size;
+        char controldigest256buffer[SceControlInfoDigest256::Size];
+        filein.read(controldigest256buffer, SceControlInfoDigest256::Size);
+        SceControlInfoDigest256 controldigest256 = SceControlInfoDigest256(controldigest256buffer);
+    }
+    filein.seekg(self_hdr.controlinfo_offset + ci_off);
+    filein.read(controlinfobuffer, SceControlInfo::Size);
+    controlinfo_hdr = SceControlInfo(controlinfobuffer);
+    ci_off += SceControlInfo::Size;
+
+    if (controlinfo_hdr.type == ControlType::NPDRM_VITA) {
+        filein.seekg(self_hdr.controlinfo_offset + ci_off);
+        ci_off += SceControlInfoDRM::Size;
+        char controlnpdrmbuffer[SceControlInfoDRM::Size];
+        filein.read(controlnpdrmbuffer, SceControlInfoDRM::Size);
+        SceControlInfoDRM controlnpdrm = SceControlInfoDRM(controlnpdrmbuffer);
+        npdrmtype = controlnpdrm.npdrm_type;
+    }
+
+    filein.seekg(self_hdr.elf_offset);
+    char dat[ElfHeader::Size];
+    filein.read(dat, ElfHeader::Size);
+    fileout.write(dat, ElfHeader::Size);
+
+    ElfHeader elf_hdr = ElfHeader(dat);
+    std::vector<ElfPhdr> elf_phdrs;
+    std::vector<SegmentInfo> segment_infos;
+    bool encrypted = false;
+    int at = ElfHeader::Size;
+
+    for (int i = 0; i < elf_hdr.e_phnum; i++) {
+        filein.seekg(self_hdr.phdr_offset + i * ElfPhdr::Size);
+        char dat[ElfPhdr::Size];
+        filein.read(dat, ElfPhdr::Size);
+        ElfPhdr phdr = ElfPhdr(dat);
+        elf_phdrs.push_back(phdr);
+        fileout.write(dat, ElfPhdr::Size);
+        at += ElfPhdr::Size;
+
+        filein.seekg(self_hdr.segment_info_offset + i * SegmentInfo::Size);
+        char segmentinfobuffer[SegmentInfo::Size];
+        filein.read(segmentinfobuffer, SegmentInfo::Size);
+        SegmentInfo segment_info = SegmentInfo(segmentinfobuffer);
+        segment_infos.push_back(segment_info);
+
+        if (segment_info.plaintext == SecureBool::NO)
+            encrypted = true;
+    }
+
+    std::vector<SceSegment> scesegs;
+
+    if (encrypted) {
+        scesegs = get_segments(filein, sce, SCE_KEYS, appinfo_hdr.sys_version, appinfo_hdr.self_type, npdrmtype, klictxt);
+    }
+
+    for (int i = 0; i < elf_hdr.e_phnum; i++) {
+        int idx;
+
+        if (!scesegs.empty())
+            idx = scesegs[i].idx;
+        else
+            idx = i;
+        if (elf_phdrs[idx].p_filesz == 0)
+            continue;
+
+        int pad_len = elf_phdrs[idx].p_offset - at;
+        if (pad_len < 0)
+            LOG_ERROR("ELF p_offset Invalid");
+
+        std::string padding;
+        for (int i = 0; i < pad_len; i++) {
+            padding += "00";
+        }
+
+        auto padding_array = string_utils::string_to_byte_array(padding);
```

Comment:
Afaik this is too inefficient. Did you port this from python? 

Did this method convert number character to raw bytes? If so you can just append '\0' and write it. Am I wrong?